packages/components/src/modifiers/hds-code-editor.ts

```diff
@@ -0,0 +1,189 @@
+import hdsDarkHighlightStyle from './hds-code-editor/highlight-styles/hds-dark-highlight-style.ts';
+
+import type { PositionalArgs, NamedArgs } from 'ember-modifier';
+import type { EditorView, ViewUpdate } from '@codemirror/view';
+import type {
+  HdsCodeEditorLanguages,
+  CodemirrorGoModule,
+  CodemirrorJsonModule,
+  CodemirrorSqlModule,
+  CodemirrorHclModule,
+  CodemirrorLanguageModule,
+} from 'src/types/hds-code-editor.types';
+
+export interface HdsCodeEditorSignature {
+  Args: {
+    Named: {
+      language?: HdsCodeEditorLanguages;
+      value?: string;
+      onInput?: (newVal: string) => void;
+      onBlur?: (editor: EditorView, event: FocusEvent) => void;
+      onSetup?: (editor: EditorView) => unknown;
+    };
+  };
+}
+
+const LOADER_HEIGHT = '164px';
+
+export default class HdsCodeEditorModifier extends Modifier<HdsCodeEditorSignature> {
+  editor!: EditorView;
+  element!: HTMLElement;
+
+  onInput: HdsCodeEditorSignature['Args']['Named']['onInput'];
+
+  observer!: IntersectionObserver;
+
+  modify(
+    element: HTMLElement,
+    positional: PositionalArgs<HdsCodeEditorSignature>,
+    named: NamedArgs<HdsCodeEditorSignature>
+  ): void {
+    assert('HdsCodeEditor must have an element', element);
+    // the intersection observer makes loading unreliable in tests
+    if (config.environment === 'test') {
+      this.setupTask.perform(element, positional, named);
+    } else {
+      this.observer = new IntersectionObserver(
+        (entries) => {
+          entries.forEach((entry) => {
+            if (entry.isIntersecting && this.setupTask.performCount === 0) {
+              this.setupTask.perform(element, positional, named);
+            }
+          });
+        },
+        {
+          rootMargin: LOADER_HEIGHT,
+        }
+      );
+
+      this.observer.observe(element);
+    }
+  }
+
+  willRemove() {
+    this.observer.disconnect();
+  }
+
+  loadLanguageTask = task(
+    { drop: true },
+    async (language?: HdsCodeEditorLanguages) => {
+      if (language === undefined) {
+        return;
+      }
+
+      let module: CodemirrorLanguageModule | null = null;
+      let languageFunction = null;
+
+      switch (language) {
+        case 'go':
+          module = await import('@codemirror/lang-go');
+          languageFunction = (module as CodemirrorGoModule).go;
+          break;
+        case 'json':
+          module = await import('@codemirror/lang-json');
+          languageFunction = (module as CodemirrorJsonModule).json;
+          break;
+        case 'sql':
+          module = await import('@codemirror/lang-sql');
+          languageFunction = (module as CodemirrorSqlModule).sql;
+          break;
+        case 'hcl':
+          module = await import('codemirror-lang-hcl');
+          languageFunction = (module as CodemirrorHclModule).hcl;
+          break;
+        default:
+          throw new Error(`Language ${language} is not supported`);
+      }
+
+      return languageFunction();
```

Comment:
suggestion: in case things go wrong above
```suggestion
      return languageFunction?.();
```